File: Assignments/4/multithreading.c
//
//  multithreading.c
//  C
//
//  Created by Yassaman Ommi on 2019-11-13.
//  Copyright © 2019 Yassaman Ommi. All rights reserved.
//

#include <stdio.h>
#include <pthread.h>
#include <math.h>

int array[50], n;

void avg()
{
    float sum = 0, average;
    
    for(int i = 0; i < n; i++)
    {
        sum = sum + array[i];
    }
    average = sum / n;
    printf("The Average Value is: %f" ,average);
}

void min()
{
    int minimum = array[0];
    
    for(int i = 1; i < n; i++)
    {
        if(minimum > array[i])
        {
            minimum = array[i];
        }
    }
    printf("\nThe Minimum Value is: %d" ,minimum);
    
}

void max()
{
    
    int maximum = array[0];
    for(int i = 1; i < n; i++)
    {
        if(maximum < array[i])
        {
            maximum = array[i];
        }
    }
    printf("\nThe Maximum Value is: %d", maximum);
}

void devi()
{
    int sum = 0, habib = 0;
    float variance, deviation, average;
    
    for(int i = 0; i < n; i++)
    {
        sum = sum + array[i];
    }
    average = sum / n;
    
    for (int i = 0; i < n; i++)
    {
        habib = habib + pow((array[i] - average), 2);
    }
    variance = habib / n;
    deviation = sqrt(variance);
    printf("\nThe Standard Deviation is: %f", deviation);
    printf("\n");
    
}

int main()
{
    printf("enter the number of your numbers: ");
    scanf("%d" ,&n);
    
    for(int i = 0; i < n; i++)
    {
        scanf("%d" ,&array[i]);
    }
    
    pthread_t t1;
    pthread_t t2;
    pthread_t t3;
    pthread_t t4;
    
    pthread_create(&t1,NULL,&avg,NULL);
    pthread_join(t1,NULL);
    
    pthread_create(&t2,NULL,&min,NULL);
    pthread_join(t2,NULL);
    
    pthread_create(&t3,NULL,&max,NULL);
    pthread_join(t3,NULL);
    
    pthread_create(&t4,NULL,&devi,NULL);
    pthread_join(t4,NULL);


}

File: Assignments/4/multithreading.py
import threading
import statistics

data = []

def avg():
    print('The Average Value is: ', statistics.mean(data))

def mini():
    print('The Minimum Value is: ', min(data))

def maxi():
    print('The Maximum Value is: ', max(data))

def devi():
    print('The Standard Deviation is: ', statistics.stdev(data))

count = input("Enter the number of your numbers: ")
for i in range(int(count)):
    data.append(int(input()))
try:
    t1 = threading.Thread(target=avg)
    t2 = threading.Thread(target=mini)
    t3 = threading.Thread(target=maxi)
    t4 = threading.Thread(target=devi)

    t1.start()
    t2.start()
    t3.start()
    t4.start()

    t1.join()
    t2.join()
    t3.join()
    t4.join()
except:
    print('Error: unable to start thread')
